{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import os, progressbar, time"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Collect protein IDs to retrieve their CDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def collect_cds(fastaFile, cdsList, outFile):\n",
    "    tmp = []\n",
    "    for seqRecord in SeqIO.parse(fastaFile, format='fasta'):\n",
    "        if seqRecord.id in cdsList :\n",
    "            tmp.append(seqRecord)\n",
    "    SeqIO.write(tmp, open(outFile, 'w'), 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "HG_fasta = '../../data/ortholog_dataset/uni_HG_orthologs.faa'\n",
    "MM_fasta = '../../data/ortholog_dataset/uni_MM_orthologs.faa'\n",
    "\n",
    "HG_cds_fasta = '/media/savvy/DATA3/savvy/Genomes/model/heterocephalus_glaber/h_glaber_cds_from_genomic.fna'\n",
    "MM_cds_fasta = '/media/savvy/DATA3/savvy/Genomes/model/mus_musculus/m_musculus_cds_from_genomic.fna'\n",
    "\n",
    "HG_IDs = [ seqRecord.id for seqRecord in SeqIO.parse(HG_fasta, format='fasta')]\n",
    "MM_IDs = [ seqRecord.id for seqRecord in SeqIO.parse(MM_fasta, format='fasta')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "collect_cds(HG_cds_fasta, HG_IDs, '../../data/ortholog_dataset/uni_HG_cds_orthologs.faa')\n",
    "collect_cds(MM_cds_fasta, MM_IDs, '../../data/ortholog_dataset/uni_MM_cds_orthologs.faa')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Collect the CDS "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "HG_cds_fasta = '../../data/ortholog_dataset/uni_HG_cds_orthologs.faa'\n",
    "MM_cds_fasta = '../../data/ortholog_dataset/uni_MM_cds_orthologs.faa'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### All proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_agg_scores = pd.read_csv('../../data/aggregation_propensity/HGMM_agg_scores.csv', sep=',') \n",
    "ortho_pairs = all_agg_scores[['proteinID_x', 'proteinID_y']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Chaperone clients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "uniprot_mapping = pd.read_csv('../../data/chaperone_clients/human_ensembl_to_uniprot.tab', sep='\\t')\n",
    "hs_mm_orthologs = pd.read_csv('../../data/chaperone_clients/HS_MM_uni_ortholog_groups.csv', sep='\\t')\n",
    "hs_mm_orthologs = hs_mm_orthologs[['proteinID_x', 'proteinID_y']]\n",
    "mm_chap_clt = hs_mm_orthologs[hs_mm_orthologs['proteinID_x'].isin(uniprot_mapping['Entry'])]['proteinID_y']\n",
    "\n",
    "chap_clt_sub = all_agg_scores[all_agg_scores['proteinID_y'].isin(list(mm_chap_clt))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1298"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(chap_clt_sub)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Container of MM and HG sequences\n",
    "HG_IDs = list(all_agg_scores['proteinID_x'])\n",
    "MM_IDs = list(all_agg_scores['proteinID_y'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "HG_seq = {}\n",
    "MM_seq = {}\n",
    "\n",
    "for seqRecord in SeqIO.parse(HG_cds_fasta, format='fasta'):\n",
    "    if seqRecord.id in HG_IDs :\n",
    "        HG_seq[seqRecord.id] = seqRecord\n",
    "        \n",
    "for seqRecord in SeqIO.parse(MM_cds_fasta, format='fasta'):\n",
    "    if seqRecord.id in MM_IDs :\n",
    "        MM_seq[seqRecord.id] = seqRecord"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MM: 9518, HG: 9522\n"
     ]
    }
   ],
   "source": [
    "print(f'MM: {len(MM_seq)}, HG: {len(HG_seq)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Check cds length / cds with non standard nucleotide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = []\n",
    "for key in MM_seq.keys() :\n",
    "    cds_id = key\n",
    "    cds = MM_seq[key].seq\n",
    "    if len(cds) % 3 == 0 :\n",
    "        if list(np.unique(cds)) != ['A', 'C', 'G', 'T']:\n",
    "            tmp.append([cds_id, 'MM', 'non_standard_nucleotide', False])\n",
    "        else: \n",
    "            tmp.append([cds_id,'MM', 'standard', True])\n",
    "    else:\n",
    "        tmp.append([cds_id, 'MM', 'non_standard_length', False])\n",
    "        \n",
    "for key in HG_seq.keys() :\n",
    "    cds_id = key\n",
    "    cds = HG_seq[key].seq\n",
    "    if len(cds) % 3 == 0 :\n",
    "        if list(np.unique(cds)) != ['A', 'C', 'G', 'T']:\n",
    "            tmp.append([cds_id, 'HG', 'non_standard_nucleotide', False])\n",
    "        else: \n",
    "            tmp.append([cds_id,'HG', 'standard', True])\n",
    "    else:\n",
    "        tmp.append([cds_id, 'HG', 'non_standard_length', False])\n",
    "\n",
    "cds_validity = pd.DataFrame(tmp, columns=['proteinID', 'organism', 'description', 'valid_cds'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18944"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cds_validity[cds_validity['valid_cds'] == True])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "HG_valids = cds_validity[(cds_validity['valid_cds'] == True) & (cds_validity['organism'] == 'HG')].rename(columns={'proteinID':'proteinID_x'})\n",
    "MM_valids = cds_validity[(cds_validity['valid_cds'] == True) & (cds_validity['organism'] == 'MM')].rename(columns={'proteinID':'proteinID_y'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "HG_valids = HG_valids.merge(ortho_pairs, on='proteinID_x')\n",
    "ortho_cds_valids = HG_valids.merge(MM_valids, on='proteinID_y')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteinID_x</th>\n",
       "      <th>organism_x</th>\n",
       "      <th>description_x</th>\n",
       "      <th>valid_cds_x</th>\n",
       "      <th>proteinID_y</th>\n",
       "      <th>organism_y</th>\n",
       "      <th>description_y</th>\n",
       "      <th>valid_cds_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E3VX36</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>P51910</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E3VX52</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>P23927</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E3VX64</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>P34884</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E3VX68</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>Q9D7M8</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E3VX70</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>P61514</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9417</th>\n",
       "      <td>G5CBQ5</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>Q3TYX3</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9418</th>\n",
       "      <td>G5CBQ6</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>Q8R361</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9419</th>\n",
       "      <td>G5CBQ8</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>Q04742</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9420</th>\n",
       "      <td>G5CBQ9</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>Q64105</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9421</th>\n",
       "      <td>G5CBR0</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>A6H5Z3</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9422 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     proteinID_x organism_x description_x  valid_cds_x proteinID_y organism_y  \\\n",
       "0         E3VX36         HG      standard         True      P51910         MM   \n",
       "1         E3VX52         HG      standard         True      P23927         MM   \n",
       "2         E3VX64         HG      standard         True      P34884         MM   \n",
       "3         E3VX68         HG      standard         True      Q9D7M8         MM   \n",
       "4         E3VX70         HG      standard         True      P61514         MM   \n",
       "...          ...        ...           ...          ...         ...        ...   \n",
       "9417      G5CBQ5         HG      standard         True      Q3TYX3         MM   \n",
       "9418      G5CBQ6         HG      standard         True      Q8R361         MM   \n",
       "9419      G5CBQ8         HG      standard         True      Q04742         MM   \n",
       "9420      G5CBQ9         HG      standard         True      Q64105         MM   \n",
       "9421      G5CBR0         HG      standard         True      A6H5Z3         MM   \n",
       "\n",
       "     description_y  valid_cds_y  \n",
       "0         standard         True  \n",
       "1         standard         True  \n",
       "2         standard         True  \n",
       "3         standard         True  \n",
       "4         standard         True  \n",
       "...            ...          ...  \n",
       "9417      standard         True  \n",
       "9418      standard         True  \n",
       "9419      standard         True  \n",
       "9420      standard         True  \n",
       "9421      standard         True  \n",
       "\n",
       "[9422 rows x 8 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ortho_cds_valids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteinID_x</th>\n",
       "      <th>organism_x</th>\n",
       "      <th>description_x</th>\n",
       "      <th>valid_cds_x</th>\n",
       "      <th>proteinID_y</th>\n",
       "      <th>organism_y</th>\n",
       "      <th>description_y</th>\n",
       "      <th>valid_cds_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E3VX36</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>P51910</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E3VX52</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>P23927</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E3VX64</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>P34884</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E3VX68</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>Q9D7M8</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E3VX70</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>P61514</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9417</th>\n",
       "      <td>G5CBQ5</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>Q3TYX3</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9418</th>\n",
       "      <td>G5CBQ6</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>Q8R361</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9419</th>\n",
       "      <td>G5CBQ8</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>Q04742</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9420</th>\n",
       "      <td>G5CBQ9</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>Q64105</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9421</th>\n",
       "      <td>G5CBR0</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>A6H5Z3</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9422 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     proteinID_x organism_x description_x  valid_cds_x proteinID_y organism_y  \\\n",
       "0         E3VX36         HG      standard         True      P51910         MM   \n",
       "1         E3VX52         HG      standard         True      P23927         MM   \n",
       "2         E3VX64         HG      standard         True      P34884         MM   \n",
       "3         E3VX68         HG      standard         True      Q9D7M8         MM   \n",
       "4         E3VX70         HG      standard         True      P61514         MM   \n",
       "...          ...        ...           ...          ...         ...        ...   \n",
       "9417      G5CBQ5         HG      standard         True      Q3TYX3         MM   \n",
       "9418      G5CBQ6         HG      standard         True      Q8R361         MM   \n",
       "9419      G5CBQ8         HG      standard         True      Q04742         MM   \n",
       "9420      G5CBQ9         HG      standard         True      Q64105         MM   \n",
       "9421      G5CBR0         HG      standard         True      A6H5Z3         MM   \n",
       "\n",
       "     description_y  valid_cds_y  \n",
       "0         standard         True  \n",
       "1         standard         True  \n",
       "2         standard         True  \n",
       "3         standard         True  \n",
       "4         standard         True  \n",
       "...            ...          ...  \n",
       "9417      standard         True  \n",
       "9418      standard         True  \n",
       "9419      standard         True  \n",
       "9420      standard         True  \n",
       "9421      standard         True  \n",
       "\n",
       "[9422 rows x 8 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### Number of valid cds pairs\n",
    "ortho_cds_valids.to_csv('../../data/mutation_tolerance/ortho_valids_cds.csv', sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteinID_x</th>\n",
       "      <th>organism_x</th>\n",
       "      <th>description_x</th>\n",
       "      <th>valid_cds_x</th>\n",
       "      <th>proteinID_y</th>\n",
       "      <th>organism_y</th>\n",
       "      <th>description_y</th>\n",
       "      <th>valid_cds_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E3VX36</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>P51910</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E3VX64</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>P34884</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>G5AJR9</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>Q8BZ60</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>G5AJS0</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>Q99PM3</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>G5AJS7</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>Q62086</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9397</th>\n",
       "      <td>G5CBL6</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>Q78XR0</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9409</th>\n",
       "      <td>G5CBN2</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>Q9Z2F6</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9412</th>\n",
       "      <td>G5CBP7</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>Q8C1M2</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9413</th>\n",
       "      <td>G5CBQ1</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>Q8K4E0</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9420</th>\n",
       "      <td>G5CBQ9</td>\n",
       "      <td>HG</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "      <td>Q64105</td>\n",
       "      <td>MM</td>\n",
       "      <td>standard</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1291 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     proteinID_x organism_x description_x  valid_cds_x proteinID_y organism_y  \\\n",
       "0         E3VX36         HG      standard         True      P51910         MM   \n",
       "2         E3VX64         HG      standard         True      P34884         MM   \n",
       "21        G5AJR9         HG      standard         True      Q8BZ60         MM   \n",
       "22        G5AJS0         HG      standard         True      Q99PM3         MM   \n",
       "26        G5AJS7         HG      standard         True      Q62086         MM   \n",
       "...          ...        ...           ...          ...         ...        ...   \n",
       "9397      G5CBL6         HG      standard         True      Q78XR0         MM   \n",
       "9409      G5CBN2         HG      standard         True      Q9Z2F6         MM   \n",
       "9412      G5CBP7         HG      standard         True      Q8C1M2         MM   \n",
       "9413      G5CBQ1         HG      standard         True      Q8K4E0         MM   \n",
       "9420      G5CBQ9         HG      standard         True      Q64105         MM   \n",
       "\n",
       "     description_y  valid_cds_y  \n",
       "0         standard         True  \n",
       "2         standard         True  \n",
       "21        standard         True  \n",
       "22        standard         True  \n",
       "26        standard         True  \n",
       "...            ...          ...  \n",
       "9397      standard         True  \n",
       "9409      standard         True  \n",
       "9412      standard         True  \n",
       "9413      standard         True  \n",
       "9420      standard         True  \n",
       "\n",
       "[1291 rows x 8 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### Number of valid cds pairs (chaperone client proteins)\n",
    "ortho_cds_valids[ortho_cds_valids['proteinID_y'].isin(mm_chap_clt)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open('../../data/mutation_tolerance/tango_directory_list/MM_chap_list', 'w')\n",
    "for ID in MM_chap_clt_stats['proteinID_y'].values:\n",
    "    f.write(f'{ID}\\n')\n",
    "f.close()    \n",
    "\n",
    "f = open('../../data/mutation_tolerance/tango_directory_list/HG_chap_list', 'w')\n",
    "for ID in HG_chap_clt_stats['proteinID_x'].values:\n",
    "    f.write(f'{ID}\\n')\n",
    "f.close()    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open('../../data/mutation_tolerance/tango_directory_list/MM_others_list', 'w')\n",
    "for ID in MM_others_stats['proteinID_y'].values:\n",
    "    f.write(f'{ID}\\n')\n",
    "f.close()    \n",
    "\n",
    "f = open('../../data/mutation_tolerance/tango_directory_list//HG_others_list', 'w')\n",
    "for ID in HG_others_stats['proteinID_x'].values:\n",
    "    f.write(f'{ID}\\n')\n",
    "f.close()    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Tango execution time\n",
    "\n",
    "- Shortest protein (Q00LT2 - length=53 aa)\n",
    "    - Time execution in Tango: 0.024s\n",
    "- Longest protein (G5C996 - length=10159 aa)\n",
    "    - Time execution in Tango: 5.116s\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Check number of mutants per protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "HG_all_count = pd.read_csv('../../data/mutation_tolerance/HG_all_mutants_counts.csv').rename(columns={'proteinID':'proteinID_x'})\n",
    "MM_all_count = pd.read_csv('../../data/mutation_tolerance/MM_all_mutants_counts.csv').rename(columns={'proteinID':'proteinID_y'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteinID_x</th>\n",
       "      <th>all_MT</th>\n",
       "      <th>unique_MT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2475</th>\n",
       "      <td>G5AYN3</td>\n",
       "      <td>14473</td>\n",
       "      <td>10007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8576</th>\n",
       "      <td>G5BYG3</td>\n",
       "      <td>14581</td>\n",
       "      <td>10015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2176</th>\n",
       "      <td>G5AX45</td>\n",
       "      <td>14788</td>\n",
       "      <td>10022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10502</th>\n",
       "      <td>G5C9E9</td>\n",
       "      <td>15058</td>\n",
       "      <td>10025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>920</th>\n",
       "      <td>G5AQ52</td>\n",
       "      <td>14518</td>\n",
       "      <td>10046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10426</th>\n",
       "      <td>G5C8Z7</td>\n",
       "      <td>50365</td>\n",
       "      <td>34644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3747</th>\n",
       "      <td>G5B601</td>\n",
       "      <td>52138</td>\n",
       "      <td>36782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7260</th>\n",
       "      <td>G5BR15</td>\n",
       "      <td>56656</td>\n",
       "      <td>39544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8958</th>\n",
       "      <td>G5C0E1</td>\n",
       "      <td>79435</td>\n",
       "      <td>55187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10476</th>\n",
       "      <td>G5C996</td>\n",
       "      <td>91441</td>\n",
       "      <td>61617</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>476 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      proteinID_x  all_MT  unique_MT\n",
       "2475       G5AYN3   14473      10007\n",
       "8576       G5BYG3   14581      10015\n",
       "2176       G5AX45   14788      10022\n",
       "10502      G5C9E9   15058      10025\n",
       "920        G5AQ52   14518      10046\n",
       "...           ...     ...        ...\n",
       "10426      G5C8Z7   50365      34644\n",
       "3747       G5B601   52138      36782\n",
       "7260       G5BR15   56656      39544\n",
       "8958       G5C0E1   79435      55187\n",
       "10476      G5C996   91441      61617\n",
       "\n",
       "[476 rows x 3 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HG_all_count[HG_all_count['unique_MT'] > 10000].sort_values('unique_MT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteinID_y</th>\n",
       "      <th>all_MT</th>\n",
       "      <th>unique_MT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1295</th>\n",
       "      <td>O88572</td>\n",
       "      <td>14527</td>\n",
       "      <td>10012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>A0A338P6S8</td>\n",
       "      <td>14743</td>\n",
       "      <td>10012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4022</th>\n",
       "      <td>Q5NCJ1</td>\n",
       "      <td>14491</td>\n",
       "      <td>10021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1404</th>\n",
       "      <td>P02468</td>\n",
       "      <td>14473</td>\n",
       "      <td>10048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1403</th>\n",
       "      <td>P02463</td>\n",
       "      <td>15031</td>\n",
       "      <td>10061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>A2AN08</td>\n",
       "      <td>46630</td>\n",
       "      <td>31953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5541</th>\n",
       "      <td>Q80W93</td>\n",
       "      <td>46396</td>\n",
       "      <td>32033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>A2ANY6</td>\n",
       "      <td>50311</td>\n",
       "      <td>34546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>561</th>\n",
       "      <td>E9Q1W3</td>\n",
       "      <td>64378</td>\n",
       "      <td>45464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8086</th>\n",
       "      <td>Q91ZU6</td>\n",
       "      <td>66547</td>\n",
       "      <td>46086</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>445 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     proteinID_y  all_MT  unique_MT\n",
       "1295      O88572   14527      10012\n",
       "65    A0A338P6S8   14743      10012\n",
       "4022      Q5NCJ1   14491      10021\n",
       "1404      P02468   14473      10048\n",
       "1403      P02463   15031      10061\n",
       "...          ...     ...        ...\n",
       "181       A2AN08   46630      31953\n",
       "5541      Q80W93   46396      32033\n",
       "183       A2ANY6   50311      34546\n",
       "561       E9Q1W3   64378      45464\n",
       "8086      Q91ZU6   66547      46086\n",
       "\n",
       "[445 rows x 3 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MM_all_count[MM_all_count['unique_MT'] > 10000].sort_values('unique_MT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "69896045"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ALL_HG = sum(HG_all_count[HG_all_count['proteinID_x'].isin(ortho_cds_valids['proteinID_x'])]['unique_MT'])\n",
    "ALL_MM = sum(MM_all_count[MM_all_count['proteinID_y'].isin(ortho_cds_valids['proteinID_y'])]['unique_MT'])\n",
    "\n",
    "ALL_HG + ALL_MM\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "HG_chap_clt_counts = pd.read_csv('../../data/mutation_tolerance/HG_chap_client_mutants_counts.csv').rename(columns={'proteinID':'proteinID_x'})\n",
    "MM_chap_clt_counts = pd.read_csv('../../data/mutation_tolerance/MM_chap_client_mutants_counts.csv').rename(columns={'proteinID':'proteinID_y'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10071232"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CHAP_CLT_HG = sum(HG_chap_clt_counts[HG_chap_clt_counts['proteinID_x'].isin(ortho_cds_valids['proteinID_x'])]['unique_MT'])\n",
    "CHAP_CLT_MM = sum(MM_chap_clt_counts[MM_chap_clt_counts['proteinID_y'].isin(ortho_cds_valids['proteinID_y'])]['unique_MT'])\n",
    "CHAP_CLT_HG + CHAP_CLT_MM"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Directory list for sbatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'HG_all_stats' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_15613/3780042585.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mHG_others_stats\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mHG_all_stats\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m~\u001b[0m\u001b[0mHG_all_stats\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'proteinID_x'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mHG_chap_clt_stats\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'proteinID_x'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mMM_others_stats\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mMM_all_stats\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m~\u001b[0m\u001b[0mMM_all_stats\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'proteinID_y'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mMM_chap_clt_stats\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'proteinID_y'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'HG_all_stats' is not defined"
     ]
    }
   ],
   "source": [
    "HG_others_stats = HG_all_stats[~HG_all_stats['proteinID_x'].isin(HG_chap_clt_stats['proteinID_x'])]\n",
    "MM_others_stats = MM_all_stats[~MM_all_stats['proteinID_y'].isin(MM_chap_clt_stats['proteinID_y'])]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Checking number of mutants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "35014973"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(HG_all_stats['unique_MT'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "34559953"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(MM_all_stats['unique_MT'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteinID_y</th>\n",
       "      <th>all_MT</th>\n",
       "      <th>unique_MT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>A2AN08</td>\n",
       "      <td>46630</td>\n",
       "      <td>31953</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   proteinID_y  all_MT  unique_MT\n",
       "12      A2AN08   46630      31953"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MM_all_stats[MM_chap_clt_stats['proteinID_y'] == 'A2AN08']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteinID_x</th>\n",
       "      <th>all_MT</th>\n",
       "      <th>unique_MT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>291</th>\n",
       "      <td>G5AWV7</td>\n",
       "      <td>40375</td>\n",
       "      <td>28247</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    proteinID_x  all_MT  unique_MT\n",
       "291      G5AWV7   40375      28247"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HG_all_stats[HG_chap_clt_stats['proteinID_x'] == 'G5AWV7']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Calculation of mutation tolerance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "MUT_RES = '/media/savvy/DATA3/savvy/project_2018/computational_mutagenesis/FINAL_RESULTS'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "  def count_final_agg_score(proteinID, MM_all_count, HG_all_count, MUT_RES):\n",
    "    if proteinID in MM_all_count['proteinID_y'].values:\n",
    "        exp = int(MM_all_count[ MM_all_count['proteinID_y'] == proteinID ]['unique_MT'].values)\n",
    "    elif proteinID in HG_all_count['proteinID_x'].values:\n",
    "        exp = int(HG_all_count[ HG_all_count['proteinID_x'] == proteinID ]['unique_MT'].values)\n",
    "    \n",
    "    if f'{proteinID}_aggregation.txt' in os.listdir(MUT_RES):\n",
    "        agg_tmp = pd.read_csv(os.path.join(MUT_RES,f'{proteinID}_aggregation.txt'), sep='\\t')\n",
    "        real = len(agg_tmp)\n",
    "            \n",
    "    if exp == real :\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "\n",
    "def calculate_mutation_metrics(agg_table, threshold):\n",
    "    agg_table['deltaMut'] = agg_table['Aggregation'] - agg_table['Aggregation'][0]\n",
    "    agg_table = agg_table[1:]\n",
    "    mutTol = len(agg_table[agg_table['deltaMut'] == 0 ]) / len(agg_table)\n",
    "    mutVul = len(agg_table[agg_table['deltaMut'] > threshold]) / len(agg_table)\n",
    "    return mutTol, mutVul\n",
    "\n",
    "\n",
    "def mutation_metrics_table(all_agg_scores, threshold):\n",
    "    tmp = []\n",
    "    bar = progressbar.ProgressBar()\n",
    "    for X, Y in bar(all_agg_scores[['proteinID_x', 'proteinID_y']].values):\n",
    "        try:\n",
    "            if f'{X}_aggregation.txt' in os.listdir(MUT_RES):\n",
    "                HG_tmp = pd.read_csv(os.path.join(MUT_RES,f'{X}_aggregation.txt'), sep='\\t')\n",
    "                MM_tmp = pd.read_csv(os.path.join(MUT_RES,f'{Y}_aggregation.txt'), sep='\\t')\n",
    "                mutTol_x, mutVul_x = calculate_mutation_metrics(HG_tmp, threshold)\n",
    "                mutTol_y, mutVul_y = calculate_mutation_metrics(MM_tmp, threshold)\n",
    "                tmp.append([X, mutTol_x, mutVul_x, Y, mutTol_y, mutVul_y])\n",
    "        except:\n",
    "            pass\n",
    "    mutation_metrics_table = pd.DataFrame(tmp, columns=['proteinID_x', 'mutTol_x', 'mutVul_x' ,'proteinID_y', 'mutTol_y', 'mutVul_y'])\n",
    "    \n",
    "    aggregation_table = all_agg_scores[['proteinID_x', 'Aggregation_x', 'length_x', 'proteinID_y', 'Aggregation_y', 'length_y']]\n",
    "    aggregation_table = aggregation_table[aggregation_table['proteinID_x'].isin(mutation_metrics_table['proteinID_x'])]\n",
    "    \n",
    "    TABLE = mutation_metrics_table.merge(aggregation_table, on=['proteinID_x', 'proteinID_y']) \n",
    "    return TABLE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100% (21651 of 21651) |##################| Elapsed Time: 0:09:13 Time:  0:09:13\n"
     ]
    }
   ],
   "source": [
    "tmp = []\n",
    "bar = progressbar.ProgressBar()\n",
    "for agg_table in bar(os.listdir(MUT_RES)):\n",
    "    protein = agg_table.split('_')[0]\n",
    "    tmp.append([protein, count_final_agg_score(protein, MM_all_count, HG_all_count, MUT_RES)])\n",
    "check_final_mutants = pd.DataFrame(tmp, columns=['proteinID', 'all_tango_run'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteinID</th>\n",
       "      <th>all_tango_run</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q7TSI0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P39061</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2AWP8</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G5AX95</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G5B286</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21646</th>\n",
       "      <td>Q61120</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21647</th>\n",
       "      <td>G5BLD0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21648</th>\n",
       "      <td>G5B8Q4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21649</th>\n",
       "      <td>G5AVI8</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21650</th>\n",
       "      <td>Q3UMR0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21637 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      proteinID  all_tango_run\n",
       "0        Q7TSI0           True\n",
       "1        P39061           True\n",
       "2        A2AWP8           True\n",
       "3        G5AX95           True\n",
       "4        G5B286           True\n",
       "...         ...            ...\n",
       "21646    Q61120           True\n",
       "21647    G5BLD0           True\n",
       "21648    G5B8Q4           True\n",
       "21649    G5AVI8           True\n",
       "21650    Q3UMR0           True\n",
       "\n",
       "[21637 rows x 2 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_final_mutants[check_final_mutants['all_tango_run'] == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutant_size = []\n",
    "for ID in check_final_mutants[check_final_mutants['all_tango_run'] == False]['proteinID']:\n",
    "    if ID in MM_all_count['proteinID_y'].values:\n",
    "        size = MM_all_count[MM_all_count['proteinID_y'] == ID]['unique_MT'].values[0]\n",
    "    elif ID in HG_all_count['proteinID_x'].values:\n",
    "        size = HG_all_count[HG_all_count['proteinID_x'] == ID]['unique_MT'].values[0]\n",
    "    mutant_size.append([ID, size])\n",
    "check_bug_mutants = pd.DataFrame(mutant_size, columns=['proteinID', 'nb_mutants'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteinID</th>\n",
       "      <th>nb_mutants</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q91ZU6</td>\n",
       "      <td>46086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G5B601</td>\n",
       "      <td>36782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B1AR51</td>\n",
       "      <td>27783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G5B5V9</td>\n",
       "      <td>30563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G5AMU8</td>\n",
       "      <td>29212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>G5C8Z7</td>\n",
       "      <td>34644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>E9Q1W3</td>\n",
       "      <td>45464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A2ANY6</td>\n",
       "      <td>34546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>G5AK10</td>\n",
       "      <td>12367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>G5C996</td>\n",
       "      <td>61617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>G5BR15</td>\n",
       "      <td>39544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>G5AKD0</td>\n",
       "      <td>16165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>G5C0E1</td>\n",
       "      <td>55187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>E9PZP8</td>\n",
       "      <td>29745</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   proteinID  nb_mutants\n",
       "0     Q91ZU6       46086\n",
       "1     G5B601       36782\n",
       "2     B1AR51       27783\n",
       "3     G5B5V9       30563\n",
       "4     G5AMU8       29212\n",
       "5     G5C8Z7       34644\n",
       "6     E9Q1W3       45464\n",
       "7     A2ANY6       34546\n",
       "8     G5AK10       12367\n",
       "9     G5C996       61617\n",
       "10    G5BR15       39544\n",
       "11    G5AKD0       16165\n",
       "12    G5C0E1       55187\n",
       "13    E9PZP8       29745"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_bug_mutants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 5.39 ms, sys: 896 µs, total: 6.28 ms\n",
      "Wall time: 5.31 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "MT_2 = mutation_metrics_table(all_agg_scores, 2)\n",
    "MT_2 = MT_2[~(MT_2['proteinID_x'].isin(check_bug_mutants['proteinID'])) | (MT_2['proteinID_y'].isin(check_bug_mutants['proteinID']))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteinID_x</th>\n",
       "      <th>mutTol_x</th>\n",
       "      <th>mutVul_x</th>\n",
       "      <th>proteinID_y</th>\n",
       "      <th>mutTol_y</th>\n",
       "      <th>mutVul_y</th>\n",
       "      <th>Aggregation_x</th>\n",
       "      <th>length_x</th>\n",
       "      <th>Aggregation_y</th>\n",
       "      <th>length_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G5B678</td>\n",
       "      <td>0.390741</td>\n",
       "      <td>0.114815</td>\n",
       "      <td>Q3UHI4</td>\n",
       "      <td>0.512232</td>\n",
       "      <td>0.094858</td>\n",
       "      <td>4.147276</td>\n",
       "      <td>261</td>\n",
       "      <td>3.305945</td>\n",
       "      <td>326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G5AQ71</td>\n",
       "      <td>0.362447</td>\n",
       "      <td>0.134984</td>\n",
       "      <td>A2CG49</td>\n",
       "      <td>0.355629</td>\n",
       "      <td>0.135774</td>\n",
       "      <td>3.033646</td>\n",
       "      <td>2400</td>\n",
       "      <td>3.007453</td>\n",
       "      <td>2964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G5C1Q7</td>\n",
       "      <td>0.199303</td>\n",
       "      <td>0.225087</td>\n",
       "      <td>A0A1W2P7S4</td>\n",
       "      <td>0.328440</td>\n",
       "      <td>0.174312</td>\n",
       "      <td>2.982352</td>\n",
       "      <td>230</td>\n",
       "      <td>1.082609</td>\n",
       "      <td>437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G5BQT4</td>\n",
       "      <td>0.256410</td>\n",
       "      <td>0.127739</td>\n",
       "      <td>Q9QYK9</td>\n",
       "      <td>0.303271</td>\n",
       "      <td>0.140187</td>\n",
       "      <td>4.179275</td>\n",
       "      <td>346</td>\n",
       "      <td>4.195598</td>\n",
       "      <td>343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G5CB95</td>\n",
       "      <td>0.499500</td>\n",
       "      <td>0.098420</td>\n",
       "      <td>Q8C120</td>\n",
       "      <td>0.462818</td>\n",
       "      <td>0.106876</td>\n",
       "      <td>2.197379</td>\n",
       "      <td>821</td>\n",
       "      <td>1.644278</td>\n",
       "      <td>878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9344</th>\n",
       "      <td>G5CAQ3</td>\n",
       "      <td>0.464617</td>\n",
       "      <td>0.088457</td>\n",
       "      <td>E9Q1A5</td>\n",
       "      <td>0.463676</td>\n",
       "      <td>0.087782</td>\n",
       "      <td>0.594638</td>\n",
       "      <td>555</td>\n",
       "      <td>0.563788</td>\n",
       "      <td>584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9345</th>\n",
       "      <td>G5AWC8</td>\n",
       "      <td>0.236297</td>\n",
       "      <td>0.111469</td>\n",
       "      <td>Q3UXZ6</td>\n",
       "      <td>0.382700</td>\n",
       "      <td>0.118829</td>\n",
       "      <td>0.408910</td>\n",
       "      <td>346</td>\n",
       "      <td>0.761354</td>\n",
       "      <td>364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9346</th>\n",
       "      <td>G5C5H8</td>\n",
       "      <td>0.509112</td>\n",
       "      <td>0.102506</td>\n",
       "      <td>Q62231</td>\n",
       "      <td>0.503127</td>\n",
       "      <td>0.104605</td>\n",
       "      <td>0.613761</td>\n",
       "      <td>284</td>\n",
       "      <td>0.619046</td>\n",
       "      <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9347</th>\n",
       "      <td>G5AR65</td>\n",
       "      <td>0.411550</td>\n",
       "      <td>0.070175</td>\n",
       "      <td>Q80VJ2</td>\n",
       "      <td>0.529825</td>\n",
       "      <td>0.056140</td>\n",
       "      <td>0.573265</td>\n",
       "      <td>223</td>\n",
       "      <td>0.547302</td>\n",
       "      <td>232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9348</th>\n",
       "      <td>G5B2K3</td>\n",
       "      <td>0.112633</td>\n",
       "      <td>0.218163</td>\n",
       "      <td>Q80X95</td>\n",
       "      <td>0.109638</td>\n",
       "      <td>0.217746</td>\n",
       "      <td>9.591840</td>\n",
       "      <td>313</td>\n",
       "      <td>9.591792</td>\n",
       "      <td>313</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9346 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     proteinID_x  mutTol_x  mutVul_x proteinID_y  mutTol_y  mutVul_y  \\\n",
       "0         G5B678  0.390741  0.114815      Q3UHI4  0.512232  0.094858   \n",
       "1         G5AQ71  0.362447  0.134984      A2CG49  0.355629  0.135774   \n",
       "2         G5C1Q7  0.199303  0.225087  A0A1W2P7S4  0.328440  0.174312   \n",
       "3         G5BQT4  0.256410  0.127739      Q9QYK9  0.303271  0.140187   \n",
       "4         G5CB95  0.499500  0.098420      Q8C120  0.462818  0.106876   \n",
       "...          ...       ...       ...         ...       ...       ...   \n",
       "9344      G5CAQ3  0.464617  0.088457      E9Q1A5  0.463676  0.087782   \n",
       "9345      G5AWC8  0.236297  0.111469      Q3UXZ6  0.382700  0.118829   \n",
       "9346      G5C5H8  0.509112  0.102506      Q62231  0.503127  0.104605   \n",
       "9347      G5AR65  0.411550  0.070175      Q80VJ2  0.529825  0.056140   \n",
       "9348      G5B2K3  0.112633  0.218163      Q80X95  0.109638  0.217746   \n",
       "\n",
       "      Aggregation_x  length_x  Aggregation_y  length_y  \n",
       "0          4.147276       261       3.305945       326  \n",
       "1          3.033646      2400       3.007453      2964  \n",
       "2          2.982352       230       1.082609       437  \n",
       "3          4.179275       346       4.195598       343  \n",
       "4          2.197379       821       1.644278       878  \n",
       "...             ...       ...            ...       ...  \n",
       "9344       0.594638       555       0.563788       584  \n",
       "9345       0.408910       346       0.761354       364  \n",
       "9346       0.613761       284       0.619046       284  \n",
       "9347       0.573265       223       0.547302       232  \n",
       "9348       9.591840       313       9.591792       313  \n",
       "\n",
       "[9346 rows x 10 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MT_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "MT_2.to_csv('../../data/mutation_tolerance/all_mt_scores.csv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__________"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data for revisions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Variations of definition in mutation tolerance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_mutation_metrics(agg_table, threshold):\n",
    "    agg_table['deltaMut'] = agg_table['Aggregation'] - agg_table['Aggregation'][0]\n",
    "    agg_table = agg_table[1:]\n",
    "    mutTol_v1 = len(agg_table[agg_table['deltaMut'] == 0 ]) / len(agg_table)\n",
    "    mutTol_v2 = len(agg_table[agg_table['deltaMut'] <= 0 ]) / len(agg_table)\n",
    "    mutTol_v3 = len(agg_table[agg_table['deltaMut'] < - threshold ]) / len(agg_table)\n",
    "    return mutTol_v1, mutTol_v2, mutTol_v3\n",
    "\n",
    "\n",
    "def mutation_metrics_table(all_agg_scores, threshold):\n",
    "    tmp = []\n",
    "    bar = progressbar.ProgressBar()\n",
    "    for X, Y in bar(all_agg_scores[['proteinID_x', 'proteinID_y']].values):\n",
    "        try:\n",
    "            if f'{X}_aggregation.txt' in os.listdir(MUT_RES):\n",
    "                HG_tmp = pd.read_csv(os.path.join(MUT_RES,f'{X}_aggregation.txt'), sep='\\t')\n",
    "                MM_tmp = pd.read_csv(os.path.join(MUT_RES,f'{Y}_aggregation.txt'), sep='\\t')\n",
    "                mutTol_v1_x, mutTol_v2_x, mutTol_v3_x = calculate_mutation_metrics(HG_tmp, threshold)\n",
    "                mutTol_v1_y, mutTol_v2_y, mutTol_v3_y = calculate_mutation_metrics(MM_tmp, threshold)\n",
    "                tmp.append([X, mutTol_v1_x, mutTol_v2_x, mutTol_v3_x, Y, mutTol_v1_y, mutTol_v2_y, mutTol_v3_y])\n",
    "        except:\n",
    "            pass\n",
    "    mutation_metrics_table = pd.DataFrame(tmp, columns=['proteinID_x', 'mutTol_v1_x', 'mutTol_v2_x', 'mutTol_v3_x' ,'proteinID_y', 'mutTol_v1_y', 'mutTol_v2_y', 'mutTol_v3_y'])\n",
    "    \n",
    "    aggregation_table = all_agg_scores[['proteinID_x', 'Aggregation_x', 'length_x', 'proteinID_y', 'Aggregation_y', 'length_y']]\n",
    "    aggregation_table = aggregation_table[aggregation_table['proteinID_x'].isin(mutation_metrics_table['proteinID_x'])]\n",
    "    \n",
    "    TABLE = mutation_metrics_table.merge(aggregation_table, on=['proteinID_x', 'proteinID_y']) \n",
    "    return TABLE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100% (9522 of 9522) |####################| Elapsed Time: 0:02:43 Time:  0:02:43\n"
     ]
    }
   ],
   "source": [
    "MT_var = mutation_metrics_table(all_agg_scores, 2)\n",
    "MT_var = MT_var[~(MT_var['proteinID_x'].isin(check_bug_mutants['proteinID'])) | (MT_var['proteinID_y'].isin(check_bug_mutants['proteinID']))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteinID_x</th>\n",
       "      <th>mutTol_v1_x</th>\n",
       "      <th>mutTol_v2_x</th>\n",
       "      <th>mutTol_v3_x</th>\n",
       "      <th>proteinID_y</th>\n",
       "      <th>mutTol_v1_y</th>\n",
       "      <th>mutTol_v2_y</th>\n",
       "      <th>mutTol_v3_y</th>\n",
       "      <th>Aggregation_x</th>\n",
       "      <th>length_x</th>\n",
       "      <th>Aggregation_y</th>\n",
       "      <th>length_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G5B678</td>\n",
       "      <td>0.390741</td>\n",
       "      <td>0.716667</td>\n",
       "      <td>0.187037</td>\n",
       "      <td>Q3UHI4</td>\n",
       "      <td>0.512232</td>\n",
       "      <td>0.729406</td>\n",
       "      <td>0.098852</td>\n",
       "      <td>4.147276</td>\n",
       "      <td>261</td>\n",
       "      <td>3.305945</td>\n",
       "      <td>326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G5AQ71</td>\n",
       "      <td>0.362447</td>\n",
       "      <td>0.651965</td>\n",
       "      <td>0.121372</td>\n",
       "      <td>A2CG49</td>\n",
       "      <td>0.355629</td>\n",
       "      <td>0.651509</td>\n",
       "      <td>0.122526</td>\n",
       "      <td>3.033646</td>\n",
       "      <td>2400</td>\n",
       "      <td>3.007453</td>\n",
       "      <td>2964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G5C1Q7</td>\n",
       "      <td>0.199303</td>\n",
       "      <td>0.574216</td>\n",
       "      <td>0.238328</td>\n",
       "      <td>A0A1W2P7S4</td>\n",
       "      <td>0.328440</td>\n",
       "      <td>0.605138</td>\n",
       "      <td>0.142018</td>\n",
       "      <td>2.982352</td>\n",
       "      <td>230</td>\n",
       "      <td>1.082609</td>\n",
       "      <td>437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G5BQT4</td>\n",
       "      <td>0.256410</td>\n",
       "      <td>0.719347</td>\n",
       "      <td>0.310490</td>\n",
       "      <td>Q9QYK9</td>\n",
       "      <td>0.303271</td>\n",
       "      <td>0.635981</td>\n",
       "      <td>0.170093</td>\n",
       "      <td>4.179275</td>\n",
       "      <td>346</td>\n",
       "      <td>4.195598</td>\n",
       "      <td>343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G5CB95</td>\n",
       "      <td>0.499500</td>\n",
       "      <td>0.711942</td>\n",
       "      <td>0.078616</td>\n",
       "      <td>Q8C120</td>\n",
       "      <td>0.462818</td>\n",
       "      <td>0.722720</td>\n",
       "      <td>0.124626</td>\n",
       "      <td>2.197379</td>\n",
       "      <td>821</td>\n",
       "      <td>1.644278</td>\n",
       "      <td>878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9344</th>\n",
       "      <td>G5CAQ3</td>\n",
       "      <td>0.464617</td>\n",
       "      <td>0.747390</td>\n",
       "      <td>0.087587</td>\n",
       "      <td>E9Q1A5</td>\n",
       "      <td>0.463676</td>\n",
       "      <td>0.716841</td>\n",
       "      <td>0.082554</td>\n",
       "      <td>0.594638</td>\n",
       "      <td>555</td>\n",
       "      <td>0.563788</td>\n",
       "      <td>584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9345</th>\n",
       "      <td>G5AWC8</td>\n",
       "      <td>0.236297</td>\n",
       "      <td>0.652234</td>\n",
       "      <td>0.137725</td>\n",
       "      <td>Q3UXZ6</td>\n",
       "      <td>0.382700</td>\n",
       "      <td>0.676278</td>\n",
       "      <td>0.093927</td>\n",
       "      <td>0.408910</td>\n",
       "      <td>346</td>\n",
       "      <td>0.761354</td>\n",
       "      <td>364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9346</th>\n",
       "      <td>G5C5H8</td>\n",
       "      <td>0.509112</td>\n",
       "      <td>0.716970</td>\n",
       "      <td>0.091686</td>\n",
       "      <td>Q62231</td>\n",
       "      <td>0.503127</td>\n",
       "      <td>0.707789</td>\n",
       "      <td>0.085276</td>\n",
       "      <td>0.613761</td>\n",
       "      <td>284</td>\n",
       "      <td>0.619046</td>\n",
       "      <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9347</th>\n",
       "      <td>G5AR65</td>\n",
       "      <td>0.411550</td>\n",
       "      <td>0.752193</td>\n",
       "      <td>0.046053</td>\n",
       "      <td>Q80VJ2</td>\n",
       "      <td>0.529825</td>\n",
       "      <td>0.755789</td>\n",
       "      <td>0.047719</td>\n",
       "      <td>0.573265</td>\n",
       "      <td>223</td>\n",
       "      <td>0.547302</td>\n",
       "      <td>232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9348</th>\n",
       "      <td>G5B2K3</td>\n",
       "      <td>0.112633</td>\n",
       "      <td>0.595637</td>\n",
       "      <td>0.258245</td>\n",
       "      <td>Q80X95</td>\n",
       "      <td>0.109638</td>\n",
       "      <td>0.596124</td>\n",
       "      <td>0.260071</td>\n",
       "      <td>9.591840</td>\n",
       "      <td>313</td>\n",
       "      <td>9.591792</td>\n",
       "      <td>313</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9346 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     proteinID_x  mutTol_v1_x  mutTol_v2_x  mutTol_v3_x proteinID_y  \\\n",
       "0         G5B678     0.390741     0.716667     0.187037      Q3UHI4   \n",
       "1         G5AQ71     0.362447     0.651965     0.121372      A2CG49   \n",
       "2         G5C1Q7     0.199303     0.574216     0.238328  A0A1W2P7S4   \n",
       "3         G5BQT4     0.256410     0.719347     0.310490      Q9QYK9   \n",
       "4         G5CB95     0.499500     0.711942     0.078616      Q8C120   \n",
       "...          ...          ...          ...          ...         ...   \n",
       "9344      G5CAQ3     0.464617     0.747390     0.087587      E9Q1A5   \n",
       "9345      G5AWC8     0.236297     0.652234     0.137725      Q3UXZ6   \n",
       "9346      G5C5H8     0.509112     0.716970     0.091686      Q62231   \n",
       "9347      G5AR65     0.411550     0.752193     0.046053      Q80VJ2   \n",
       "9348      G5B2K3     0.112633     0.595637     0.258245      Q80X95   \n",
       "\n",
       "      mutTol_v1_y  mutTol_v2_y  mutTol_v3_y  Aggregation_x  length_x  \\\n",
       "0        0.512232     0.729406     0.098852       4.147276       261   \n",
       "1        0.355629     0.651509     0.122526       3.033646      2400   \n",
       "2        0.328440     0.605138     0.142018       2.982352       230   \n",
       "3        0.303271     0.635981     0.170093       4.179275       346   \n",
       "4        0.462818     0.722720     0.124626       2.197379       821   \n",
       "...           ...          ...          ...            ...       ...   \n",
       "9344     0.463676     0.716841     0.082554       0.594638       555   \n",
       "9345     0.382700     0.676278     0.093927       0.408910       346   \n",
       "9346     0.503127     0.707789     0.085276       0.613761       284   \n",
       "9347     0.529825     0.755789     0.047719       0.573265       223   \n",
       "9348     0.109638     0.596124     0.260071       9.591840       313   \n",
       "\n",
       "      Aggregation_y  length_y  \n",
       "0          3.305945       326  \n",
       "1          3.007453      2964  \n",
       "2          1.082609       437  \n",
       "3          4.195598       343  \n",
       "4          1.644278       878  \n",
       "...             ...       ...  \n",
       "9344       0.563788       584  \n",
       "9345       0.761354       364  \n",
       "9346       0.619046       284  \n",
       "9347       0.547302       232  \n",
       "9348       9.591792       313  \n",
       "\n",
       "[9346 rows x 12 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MT_var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "MT_var.to_csv('../../data/mutation_tolerance/all_mt_scores_for_revisions.csv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "52454c618a4204657519cbab42273bb942d0a30f08c27de48ae386072d003f99"
  },
  "kernelspec": {
   "display_name": "Python 3.8.10 64-bit ('bioinfo_3.8': conda)",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
